fuzzers/005-tilegrid/generate_full.py

```diff
@@ -352,16 +352,18 @@ def propagate_IOI_SING(database, tiles_by_grid):
                 database[tile]['grid_x'], database[tile]['grid_y'] - 1)]
             seen_iois.add(tile)
 
-            if '_SING' in database[tile]['type']:
+            if 'IOI3_SING' in database[tile]['type']:
                 break
 
-            if 'CLB_IO_CLK' in database[tile]['bits']:
-                assert bits['baseaddr'] == database[tile]['bits'][
-                    'CLB_IO_CLK']['baseaddr']
-                assert bits['frames'] == database[tile]['bits']['CLB_IO_CLK'][
-                    'frames']
-                assert bits['words'] == database[tile]['bits']['CLB_IO_CLK'][
-                    'words']
+        if 'CLB_IO_CLK' in database[tile]['bits']:
```

Comment:
This indent should not have collapsed?  You have removed the assertion check in the loop?

Reply:
Hmm, you are correct about the indentation level, but I don't get why we want to run the assertion checks for all tiles along the Y diagonal, e.g.
<pre>
LIOI3_X0Y123
NULL_X1Y129
HCLK_IOI3_X1Y130
</pre>

For IOI I get:
<pre>
  File "../generate_full.py", line 365, in propagate_IOI_SING
    database[tile]['bits']['CLB_IO_CLK']['frames'])
AssertionError: HCLK_IOI3_X113Y26:42 == 30
</pre>

My question here should be whether the HCLK_IOI3 tiles that are being calculated in this PR actually have 30 frames instead of 42? I did the same as for HCLK_CMT, but I am not sure where to get this information from.

BTW. The same goes for IOB and HCLK_IOBs once we want to calculate the base address for it:
<pre>
NULL_X0Y77
HCLK_IOB_X0Y78
LIOB33_X0Y75
</pre>

Reply:
Frames should be <= whatever is in https://github.com/SymbiFlow/prjxray-db/blob/master/artix7/xc7a35tcpg236-1.yaml

Frames are determined at https://github.com/SymbiFlow/prjxray/pull/915/files#diff-558ab8b139f412dbaddc7136214af59dL80

> tiles along the Y diagonal, e.g.

This loop is running in a column, not in the diagonal.  Also this function shouldn't "see" the HCLK_IOI3 tile at all, instead just see [LR]IOI.  It is intended to operate between the IOI_SING types, so NULL etc shouldn't be seen.

Reply:
I wrote diagonal but had actually column in mind. Then I guess this function needs to be revisited for ioi and iob as well because hclk_iob is in the same column as iobs and hclk_ioi with iois.

Reply:
The original code was only intended to run between the IOIxx_SING and the other IOIxx_SING, with IOIxx in between.  It is unclear what is happening now.

Reply:
I guess that this has been in there from the begining, but since the assertions were only checked if CLB_IO_CLK was not empty there was no problem. Now it's not empty for HCLK_IOI and this became a problem. It will be the same case for IOBs once HCLK_IOB has a base address. Whether it will is a separate issue.